File: Coursica/FiltersViewController.m
//
//  FiltersViewController.m
//  Coursica
//
//  Created by Regan Bell on 11/25/14.
//  Matthew Beatty
//  Copyright (c) 2014 Prestige Worldwide. All rights reserved.
//

#import "FiltersViewController.h"
#import <QuartzCore/QuartzCore.h>
#import "DoubleSliderView.h"
#import "QReport.h"

#define CoursicaBlue [UIColor colorWithRed:31/255.0 green:148/255.0 blue:255/255.0 alpha:1.0]
#define UnselectedGray [UIColor colorWithRed:217/255.0 green:215/255.0 blue:215/255.0 alpha:1.0]

@interface FiltersViewController ()

@property (nonatomic, strong) IBOutletCollection(UIButton) NSArray *genEdButtons;
@property (nonatomic, strong) IBOutletCollection(UILabel) NSArray *genEdLabels;
@property (nonatomic, strong) IBOutletCollection(UIView) NSArray *cards;
@property (nonatomic, strong) IBOutletCollection(UIImageView) NSArray *genEdImageViews;

@property (nonatomic, strong) IBOutletCollection(UIButton) NSArray *termBarButtons;
@property (nonatomic, strong) IBOutletCollection(UILabel) NSArray *termBarLabels;

@property (nonatomic, strong) IBOutletCollection(UIButton) NSArray *gradBarButtons;
@property (nonatomic, strong) IBOutletCollection(UILabel) NSArray *gradBarLabels;

@property (weak, nonatomic) IBOutlet UIView *termBarView;
@property (weak, nonatomic) IBOutlet UIView *genEdBarView;

@property (weak, nonatomic) NMRangeSlider *qOverallSlider;
@property (weak, nonatomic) NMRangeSlider *qWorkloadSlider;
@property (weak, nonatomic) NMRangeSlider *enrollmentSlider;

@property (weak, nonatomic) IBOutlet UIScrollView *scrollView;

@property (assign) NSInteger selectedTermIndex;
@property (assign) NSInteger selectedGradIndex;

@property (weak, nonatomic) IBOutlet UIButton *filterCoursesButton;

@end

@implementation FiltersViewController

- (NSPredicate *)filters {
    
    NSMutableArray *predicates = [NSMutableArray new];

    switch (self.selectedGradIndex) {
        case 0:
            [predicates addObject:[NSPredicate predicateWithFormat:@"graduate = %@", [NSNumber numberWithBool:NO]]];
            break;
        case 1:
            [predicates addObject:[NSPredicate predicateWithFormat:@"graduate = %@", [NSNumber numberWithBool:YES]]];
        default:
            break;
    }
    
    [predicates addObject:[NSPredicate predicateWithFormat:@"overall >= %f", self.qOverallSlider.lowerValue]];
    [predicates addObject:[NSPredicate predicateWithFormat:@"overall <= %f", self.qOverallSlider.upperValue]];

    [predicates addObject:[NSPredicate predicateWithFormat:@"workload >= %f", self.qWorkloadSlider.lowerValue]];
    [predicates addObject:[NSPredicate predicateWithFormat:@"workload <= %f", self.qWorkloadSlider.upperValue]];

    [predicates addObject:[NSPredicate predicateWithFormat:@"enrollment >= %d", (int)self.enrollmentSlider.lowerValue]];
    
    int upperEnrollment = (int)self.enrollmentSlider.upperValue;
    if (upperEnrollment < 250) {
        [predicates addObject:[NSPredicate predicateWithFormat:@"enrollment <= %d", (int)self.enrollmentSlider.upperValue]];
    }
    
    switch (self.selectedTermIndex) {
        case 0:
            [predicates addObject:[NSPredicate predicateWithFormat:@"term = %@", @"FALL"]];
            break;
        case 1:
            [predicates addObject:[NSPredicate predicateWithFormat:@"term = %@", @"SPRING"]];
        default:
            break;
    }

//    for (UIButton *button in self.genEdButtons) {
//
//        if (button.selected) {
//            NSNumber *index = [NSNumber numberWithInteger:button.tag + 1];
//            [predicates addObject:[NSPredicate predicateWithFormat:@"genEdOne = %@ OR genEdTwo = %@", index, index]];
//        }
//    }

    return [NSCompoundPredicate andPredicateWithSubpredicates:predicates];
}

- (IBAction)filterCoursesButtonPressed:(id)sender {
    
    [self.delegate filtersDidChange];
}

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    for (UIImageView *genEdImageView in self.genEdImageViews) {
        
        genEdImageView.image = [genEdImageView.image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        genEdImageView.tintColor = UnselectedGray;
    }
    
    for (UIView *card in self.cards) {
        
        card.layer.cornerRadius = 4.0f;
        card.clipsToBounds = YES;
    }
    
    self.filterCoursesButton.layer.cornerRadius = 4.0f;
    self.filterCoursesButton.clipsToBounds = YES;
    
    int i = 0;
    for (UIButton *genEdButton in self.genEdButtons) {
        [genEdButton addTarget:self action:@selector(genEdButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
        genEdButton.tag = i;
        i++;
    }
    
    int j = 0;
    for (UIButton *termButton in self.termBarButtons) {
        [termButton addTarget:self action:@selector(termButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
        termButton.tag = j;
        j++;
    }
    
    int k = 0;
    for (UIButton *gradButton in self.gradBarButtons) {
        [gradButton addTarget:self action:@selector(gradButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
        gradButton.tag = k;
        k++;
    }
    
    self.scrollView.keyboardDismissMode = UIScrollViewKeyboardDismissModeOnDrag;
    
    UITapGestureRecognizer *recognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTap:)];
    [self.scrollView addGestureRecognizer:recognizer];
    
    [self selectButton:self.termBarButtons.lastObject inArray:self.termBarLabels];
    [self selectButton:self.gradBarButtons.lastObject inArray:self.gradBarLabels];

    [self.scrollView addSubview:self.genEdBarView];
    [self configureSliders];
}

- (void)handleTap:(UITapGestureRecognizer*)recognizer {
    
    [self.delegate keyboardShouldDismiss];
}

#pragma mark - Button Press Methods

- (void)selectButton:(UIButton*)button inArray:(NSArray*)array {
    
    button.selected = !button.selected;
    UILabel *label = array[button.tag];
    UIColor *newColor = (button.selected) ? CoursicaBlue : UnselectedGray;
    label.textColor = newColor;
}

- (void)gradButtonPressed:(UIButton*)senderButton {
    
    [self.delegate keyboardShouldDismiss];
    self.selectedGradIndex = senderButton.tag;
    [self selectButton:senderButton inArray:self.gradBarLabels];
    for (UIButton *button in self.gradBarButtons) {
        if (button.selected && button.tag != senderButton.tag) {
            [self selectButton:button inArray:self.gradBarLabels];
        }
    }
}

- (void)termButtonPressed:(UIButton*)senderButton {
    
    [self.delegate keyboardShouldDismiss];
    self.selectedTermIndex = senderButton.tag;
    [self selectButton:senderButton inArray:self.termBarLabels];
    for (UIButton *button in self.termBarButtons) {
        if (button.selected && button.tag != senderButton.tag) {
            [self selectButton:button inArray:self.termBarLabels];
        }
    }
}

- (void)genEdButtonPressed:(UIButton*)senderButton {
    
    [self.delegate keyboardShouldDismiss];
    senderButton.selected = !senderButton.selected;
    
    UIImageView *iconView = self.genEdImageViews[senderButton.tag];
    UILabel *label = self.genEdLabels[senderButton.tag];
    
    UIColor *newColor = (senderButton.selected) ? CoursicaBlue : UnselectedGray;
    [UIView animateWithDuration:0 animations:^{
        label.textColor = newColor;
        iconView.tintColor = newColor;
    }];
}
//
//- (IBAction)applyFiltersButtonPressed:(id)sender {
//    
//    [self.delegate dismissFiltersViewController];
//}
//
//- (IBAction)cancelButtonPressed:(id)sender {
//    
//    [self.delegate dismissFiltersViewController];
//}

#pragma mark - Programmatic Sliders

- (DoubleSliderView*)configureSliderWithTitle:(NSString*)title font:(UIFont*)font textColor:(UIColor*)textColor {
    
    DoubleSliderView *sliderView = [[DoubleSliderView alloc] initWithTitle:title font:font textColor:textColor];
    [self.scrollView addSubview:sliderView];
    
    [self.scrollView addConstraint:[NSLayoutConstraint constraintWithItem:sliderView attribute:NSLayoutAttributeLeading relatedBy:NSLayoutRelationEqual toItem:self.genEdBarView attribute:NSLayoutAttributeLeft multiplier:1 constant:18]];
    [self.scrollView addConstraint:[NSLayoutConstraint constraintWithItem:sliderView attribute:NSLayoutAttributeTrailing relatedBy:NSLayoutRelationEqual toItem:self.genEdBarView attribute:NSLayoutAttributeRight multiplier:1 constant:-18]];
    [sliderView addConstraint:[NSLayoutConstraint constraintWithItem:sliderView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:52.0]];
    
    return sliderView;
}

- (void) configureSliders
{
    UIFont *font = [UIFont fontWithName:@"AvenirNext-Medium" size:14.0];
    UIColor *textColor = [UIColor colorWithWhite:155/255.0 alpha:1.0];
    
    DoubleSliderView *overall = [self configureSliderWithTitle:@"Overall Q Score" font:font textColor:textColor];
    overall.shouldFormatForFloatValue = YES;
    [overall valueChanged:overall.slider];
    self.qOverallSlider = overall.slider;
    
    [self.scrollView addConstraint:[NSLayoutConstraint constraintWithItem:overall attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:self.genEdBarView attribute:NSLayoutAttributeBottom multiplier:1 constant:16]];
    
    DoubleSliderView *workload = [self configureSliderWithTitle:@"Workload" font:font textColor:textColor];
    workload.shouldFormatForFloatValue = YES;
    [workload valueChanged:workload.slider];
    self.qWorkloadSlider = workload.slider;
    
    [self.scrollView addConstraint:[NSLayoutConstraint constraintWithItem:workload attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:overall attribute:NSLayoutAttributeBottom multiplier:1 constant:16]];
    
    DoubleSliderView *enrollment = [self configureSliderWithTitle:@"Enrollment" font:font textColor:textColor];
    self.enrollmentSlider = enrollment.slider;
    enrollment.shouldFormatForFloatValue = NO;
    self.enrollmentSlider.minimumValue = 1;
    self.enrollmentSlider.maximumValue = 250;
    self.enrollmentSlider.upperValue = 250;
    [enrollment valueChanged:self.enrollmentSlider];
    
    [self.scrollView addConstraint:[NSLayoutConstraint constraintWithItem:enrollment attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:workload attribute:NSLayoutAttributeBottom multiplier:1 constant:16]];
}

@end

File: Coursica/FiltersViewController.swift
//
//  FiltersViewController.swift
//  Coursica
//
//  Created by Regan Bell on 8/5/15.
//  Copyright (c) 2015 Prestige Worldwide. All rights reserved.
//

import Cartography
import RealmSwift

let unselectedGray = UIColor(white: 217/255.0, alpha: 1)

protocol FiltersViewControllerDelegate {
    
    func filtersDidChange()
    func keyboardShouldDismiss()
}

class FiltersViewController: UIViewController {

    @IBOutlet var genEdButtons: [UIButton]!
    @IBOutlet var genEdLabels: [UILabel]!
    @IBOutlet var cards: [UIView]!
    @IBOutlet var genEdImageViews: [UIImageView]!
    @IBOutlet var termBarButtons: [UIButton]!
    @IBOutlet var termBarLabels: [UILabel]!
    @IBOutlet var gradBarButtons: [UIButton]!
    @IBOutlet var gradBarLabels: [UILabel]!
    
    @IBOutlet var termBarView: UIView!
    @IBOutlet var genEdBarView: UIView!
    
    @IBOutlet var scrollView: UIScrollView!
    
    var qOverallSlider: NMRangeSlider!
    var qWorkloadSlider: NMRangeSlider!
    var enrollmentSlider: NMRangeSlider!
    
    var selectedTermIndex = 0
    var selectedGradIndex = 0
    
    @IBOutlet var filterCoursesButton: UIButton!
    
    var delegate: FiltersViewControllerDelegate?
    var filters: NSPredicate {
        let predicateOptions: [NSPredicate?] = [gradPredicate(), termPredicate(), overallPredicate(), workloadPredicate(), enrollmentPredicate(), genEdPredicate()]
        var filters: [NSPredicate] = []
        for predicateOption in predicateOptions {
            if let predicate = predicateOption {
                filters.append(predicate)
            }
        }
        return NSCompoundPredicate.andPredicateWithSubpredicates(filters)
    }
    
    func gradPredicate() -> NSPredicate? {
        switch selectedGradIndex {
        case 0:  return NSPredicate(format: "graduate = %@", NSNumber(bool: false))
        case 1:  return NSPredicate(format: "graduate = %@", NSNumber(bool: true))
        default: return nil
        }
    }
    
    func termPredicate() -> NSPredicate? {
        switch selectedTermIndex {
        case 0:  return NSPredicate(format: "term = %@", "FALL")
        case 1:  return NSPredicate(format: "term = %@", "SPRING")
        default: return nil
        }
    }
    
    func overallPredicate() -> NSPredicate? {
        if qOverallSlider.upperValue > 4.9 && qOverallSlider.lowerValue < 1.1 {
            return nil
        } else {
            return NSPredicate(format: "overall >= %f AND overall <= %f", qOverallSlider.lowerValue, qOverallSlider.upperValue)
        }
    }
    
    func workloadPredicate() -> NSPredicate? {
        if qWorkloadSlider.upperValue > 4.9 && qWorkloadSlider.lowerValue < 1.1 {
            return nil
        } else {
            return NSPredicate(format: "workload >= %f AND workload <= %f", qWorkloadSlider.lowerValue, qWorkloadSlider.upperValue)
        }
    }
    
    func enrollmentPredicate() -> NSPredicate? {
        let lowerPredicate = NSPredicate(format: "enrollment >= %d", Int(enrollmentSlider.lowerValue))
        if Int(enrollmentSlider.upperValue) < 250 {
            let upperPredicate = NSPredicate(format: "enrollment <= %d", Int(enrollmentSlider.upperValue))
            return NSCompoundPredicate.andPredicateWithSubpredicates([lowerPredicate, upperPredicate])
        } else {
            return lowerPredicate
        }
    }
    
    func genEdPredicate() -> NSPredicate? {
        var selectedGenEdButton: UIButton? = nil
        for button in genEdButtons {
            if button.selected == true {
                selectedGenEdButton = button
            }
        }
        if let selected = selectedGenEdButton {
            
            let genEds = ["Aesthetic and Interpretive Understanding",
                          "Culture and Belief",
                          "Empirical and Mathematical Reasoning",
                          "Ethical Reasoning",
                          "Science of Living Systems",
                          "Science of the Physical Universe",
                          "Societies of the World",
                          "United States in the World",
                          "Study of the Past"]
            
            return NSPredicate(format: "ANY genEds.name = %@", genEds[selected.tag])
        } else {
            return nil
        }
    }
    
    @IBAction func filterCoursesButtonPressed(button: UIButton) {
        delegate?.filtersDidChange()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        for imageView in genEdImageViews {
            imageView.image = imageView.image?.imageWithRenderingMode(.AlwaysTemplate)
            imageView.tintColor = unselectedGray
        }
        for card in cards {
            card.layer.cornerRadius = 4
            card.clipsToBounds = true
        }
        filterCoursesButton.layer.cornerRadius = 4
        filterCoursesButton.clipsToBounds = true
        let pairs: [([UIButton], Selector)] = [(genEdButtons, "genEdButtonPressed:"), (termBarButtons, "termButtonPressed:"), (gradBarButtons, "gradButtonPressed:")]
        for pair in pairs {
            let (buttons, selector) = pair
            for (index, button) in enumerate(buttons) {
                button.tag = index
                button.addTarget(self, action: selector, forControlEvents: UIControlEvents.TouchUpInside)
            }
        }
        scrollView.keyboardDismissMode = .OnDrag
        scrollView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: "handleTap:"))
        scrollView.addSubview(genEdBarView)
        
        selectButton(termBarButtons.last!, labels: termBarLabels, buttons: termBarButtons)
        selectButton(gradBarButtons.last!, labels: gradBarLabels, buttons: gradBarButtons)
        layoutSliders()
    }
    
    func handleTap(recognizer: UITapGestureRecognizer) {
        delegate?.keyboardShouldDismiss()
    }
    
    func selectButton(selectedButton: UIButton, labels: [UILabel], buttons: [UIButton]) {
        delegate?.keyboardShouldDismiss()
        selectedButton.selected = !selectedButton.selected
        labels[selectedButton.tag].textColor = selectedButton.selected ? coursicaBlue : unselectedGray
        for button in buttons {
            if button.selected && button.tag != selectedButton.tag {
                button.selected = false
                labels[button.tag].textColor = unselectedGray
            }
        }
    }
    
    func gradButtonPressed(button: UIButton) {
        selectedGradIndex = button.tag
        selectButton(button, labels: gradBarLabels, buttons: gradBarButtons)
    }
    
    func termButtonPressed(button: UIButton) {
        selectedTermIndex = button.tag
        selectButton(button, labels: termBarLabels, buttons: termBarButtons)
    }
    
    func genEdButtonPressed(pressedButton: UIButton) {
        delegate?.keyboardShouldDismiss()
        let selected = !pressedButton.selected
        for button in genEdButtons {
            button.selected == false
            genEdImageViews[button.tag].tintColor = unselectedGray
            genEdLabels[button.tag].textColor = unselectedGray
        }
        pressedButton.selected = selected
        let newColor = selected ? coursicaBlue : unselectedGray
        genEdLabels[pressedButton.tag].textColor = newColor
        genEdImageViews[pressedButton.tag].tintColor = newColor
    }
    
    func layoutSliderWithTitle(title: String) -> DoubleSliderView {
        let font = UIFont(name: "AvenirNext-Medium", size: 14)
        let textColor = UIColor(white: 155/255.0, alpha: 1)
        let sliderView = DoubleSliderView(title: title, font: font, textColor: textColor)
        scrollView.addSubview(sliderView)
        constrain(sliderView, genEdBarView, {slider, genEdBar in
            slider.left == genEdBar.left + 18
            slider.right == genEdBar.right - 18
            slider.height == 52
        })
        return sliderView
    }
    
    func layoutSliders() {
        let overallSliderView = layoutSliderWithTitle("Overall Q Score")
        overallSliderView.shouldFormatForFloatValue = true
        overallSliderView.valueChanged(overallSliderView.slider)
        qOverallSlider = overallSliderView.slider
        constrain(overallSliderView, genEdBarView, {overall, genEdBar in
            overall.top == genEdBar.bottom + 16
        })
        
        let workloadSliderView = layoutSliderWithTitle("Workload")
        workloadSliderView.shouldFormatForFloatValue = true
        workloadSliderView.valueChanged(workloadSliderView.slider)
        qWorkloadSlider = workloadSliderView.slider
        constrain(workloadSliderView, overallSliderView, {workload, overall in
            workload.top == overall.bottom + 16
        })
        
        let enrollmentSliderView = layoutSliderWithTitle("Enrollment")
        enrollmentSliderView.shouldFormatForFloatValue = false
        enrollmentSlider = enrollmentSliderView.slider
        enrollmentSlider.minimumValue = 1
        enrollmentSlider.maximumValue = 250
        enrollmentSlider.upperValue = 250
        enrollmentSliderView.valueChanged(enrollmentSliderView.slider)
        constrain(enrollmentSliderView, workloadSliderView, {enrollment, workload in
            enrollment.top == workload.bottom + 16
        })
    }
}
